sklearn/manifold/t_sne.py

```diff
@@ -48,15 +52,55 @@ def _joint_probabilities(distances, desired_perplexity, verbose):
     # Compute conditional probabilities such that they approximately match
     # the desired perplexity
     conditional_P = _utils._binary_search_perplexity(
-        distances, desired_perplexity, verbose)
+        distances.astype(np.float), desired_perplexity, verbose)
```

Comment:
This will trigger a memory copy even if `distances` already has the right type. Please use `sklearn.utils.fixes.astype` and use it with `astype(distances, np.float, copy=False)`. Also I would rather use an explicit precision with either `np.float32` or `np.float64` or better a dtype that reflect the precision of the input data instead of an arbitrary yet platform dependent `np.float` dtype.